ONSdigital/census-rh-service
# Cr 555 firestore backoff

# Motivation and Context
This change adds support to RH so that it is tolerant of Firestore being overloaded. It will now exponentially backoff and retry the transaction multiple times. This means that the throughput of an RH instance should be just below the maximum possible and with no transactions going to the dead letter queue.

# What has changed
- Firestore code identifies Firestore contention exceptions.
- Data store code uses Spring @Retryable annotation to implement exponential backoffs.
- Added unit test.

# How to test?
- Run unit tests.
- Run RH under with empty collections a heavy enough load to provoke contention errors. Then search the log file for ': Transaction' (See Readme.md for details of what is logged)

# Links
https://collaborate2.ons.gov.uk/jira/browse/CR-555#
